common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -117,76 +119,445 @@ public Collation(
+      }
+
+      protected static final int IMPLEMENTATION_PROVIDER_OFFSET = 29;
+      protected static final int IMPLEMENTATION_PROVIDER_MASK = 0b11;
+      protected static final int CASE_SENSITIVITY_OFFSET = 28;
+      protected static final int CASE_SENSITIVITY_MASK = 0b1;
+      protected static final int ACCENT_SENSITIVITY_OFFSET = 27;
+      protected static final int ACCENT_SENSITIVITY_MASK = 0b1;
+      protected static final int CASE_CONVERSION_OFFSET = 21;
+      protected static final int CASE_CONVERSION_MASK = 0b11;
+      protected static final int LOCALE_OFFSET = 0;
+      protected static final int LOCALE_MASK = 0x0FFF;
+
+      protected static final int INDETERMINATE_COLLATION_ID =
+        ImplementationProvider.INDETERMINATE.ordinal() << IMPLEMENTATION_PROVIDER_OFFSET;
+
+      protected final CaseSensitivity caseSensitivity;
+      protected final AccentSensitivity accentSensitivity;
+      protected final CaseConversion caseConversion;
+      protected final String locale;
+      protected final int collationId;
+
+      protected CollationSpec(
+          String locale,
+          CaseSensitivity caseSensitivity,
+          AccentSensitivity accentSensitivity,
+          CaseConversion caseConversion) {
+        this.locale = locale;
+        this.caseSensitivity = caseSensitivity;
+        this.accentSensitivity = accentSensitivity;
+        this.caseConversion = caseConversion;
+        this.collationId = getCollationId();
+      }
+
+      private static final Map<Integer, Collation> collationMap = new ConcurrentHashMap<>();
+
+      public static Collation fetchCollation(int collationId) throws SparkException {
+        if (collationId == UTF8_BINARY_COLLATION_ID) {
+          return CollationSpecUTF8Binary.UTF8_BINARY_COLLATION;
+        } else if (collationMap.containsKey(collationId)) {
+          return collationMap.get(collationId);
+        } else {
+          CollationSpec spec;
+          int implementationProviderOrdinal =
+            (collationId >> IMPLEMENTATION_PROVIDER_OFFSET) & IMPLEMENTATION_PROVIDER_MASK;
+          if (implementationProviderOrdinal >= ImplementationProvider.values().length) {
+            throw SparkException.internalError("Invalid collation implementation provider");
+          } else {
+            ImplementationProvider implementationProvider = ImplementationProvider.values()[
+              implementationProviderOrdinal];
+            if (implementationProvider == ImplementationProvider.UTF8_BINARY) {
+              spec = CollationSpecUTF8Binary.fromCollationId(collationId);
+            } else if (implementationProvider == ImplementationProvider.ICU) {
+              spec = CollationSpecICU.fromCollationId(collationId);
+            } else {
+              throw SparkException.internalError("Cannot instantiate indeterminate collation");
+            }
+            Collation collation = spec.buildCollation();
+            collationMap.put(collationId, collation);
```

Comment:
Just a note - we are never removing items from this hashmap. This means that if someone goes through all collations there will be non-trivial memory impact on the system.

This is probably not a concern since I can't imagine why customer would use 10k+ different collations but may be a way to DOS the system. i.e. with given query pattern you may get system to OOM.

@cloud-fan - does Spark in general worries about such things?

I am fine with this going as is btw.

Reply:
For now (looking at changes of this PR) it's not that big deal since we're supporting ~200 locales x 2 case sensitivity x 2 accent sensitivity x 3 case conversions. So if user wants to use all available collations (highly unlikely, user will probably use several different collations on average) it will be < 2000 entries in the map.

The idea with map (cache) approach was that we don't want to hardcode all the collations into the table which was the previous approach. This will make a considerable difference when we add support for punctuation sensitivity, first letter preference and space trimming; this will grow number of available collations to ~40k. I think it would make sense to worry about caching strategies after these additions.

Reply:
is quiet map eviction a reasonable solution? i.e. just limit map size using some LRU policy

Reply:
Resolving this discussion since maximum size of the map is quite small with current scope of changes (<200 locales x 2 case sensitivity x 2 accent sensitivity ~ 800 entries total). Let's revisit this when we expand the scope to space trimming, punctuation sensitivity and first letter preference.